scarb/src/core/registry/client/http.rs

```diff
@@ -146,12 +151,77 @@ impl<'c> RegistryClient for HttpRegistryClient<'c> {
     }
 
     async fn supports_publish(&self) -> Result<bool> {
-        // TODO(mkaput): Publishing to HTTP registries is not implemented yet.
-        Ok(false)
+        let index_config = self.index_config.load().await?;
+        match index_config.upload {
+            Some(_) => Ok(true),
+            None => Ok(false),
+        }
     }
 
-    async fn publish(&self, _package: Package, _tarball: FileLockGuard) -> Result<()> {
-        todo!("Publishing to HTTP registries is not implemented yet.")
+    async fn publish(&self, package: Package, tarball: FileLockGuard) -> Result<RegistryUpload> {
+        let auth_token = env::var("SCARB_REGISTRY_AUTH_TOKEN")
+            .map_err(|_| anyhow!("missing authentication token"))?;
+
+        let path = tarball.path().to_owned();
+        // we need to drop, because windows file locking is very strict
+        drop(tarball);
+
+        ensure!(
+            Path::new(&path).exists(),
+            "cannot upload package - file does not exist at path: {}",
+            &path
+        );
+
+        let file = TokioFile::open(&path).await?;
+        let metadata = file.metadata().await?;
+        ensure!(
+            metadata.len() < 5 * 1024 * 1024,
+            "package cannot be larger than `5` MB: found `{}`",
+            &metadata.len() / 1024 / 1024
+        );
+
+        let index_config = self.index_config.load().await?;
+
+        let file_part = Part::stream(Body::from(file))
+            .file_name(format!("{}_{}", &package.id.name, &package.id.version));
+        let form = Form::new().part("file", file_part);
+
+        let response = self
+            .config
+            .online_http()?
+            .post(
+                index_config
+                    .upload
+                    .clone()
+                    .ok_or_else(|| anyhow!("upload URL is missing"))?,
+            )
+            .header(AUTHORIZATION, auth_token)
+            .multipart(form)
+            .timeout(Duration::from_secs(60))
```

Comment:
https://brr.fyi/posts/engineering-for-slow-internet - consider this

Reply:
It takes ~ 60 seconds to upload 5MB file with 700Kb/s internet connection
EDGE connectivity standard (2.5G) approaches speeds up to 1Mb/s (less than that on average for a typical connection) - this is second slowest standard, right above GPRS (GSM) with 30 – 80 kb/s.
From what I gather slowest internet speeds are mostly in Africa, with lowest probably in the region of Yemen with ~700Kb/s

With all this context an educated guess was to do 60s. We can raise it to eg 120s if you prefer so :+1:

Reply:
I would rather have no timeout at all and instead show a message that upload is in progress but it's running slow after some amount of time

Reply:
yeah, that TODO was from times I wasn't aware of slow internet problems

this is not urgent

Reply:
great, created an issue https://github.com/software-mansion/scarb/issues/1580
and removed the timeout in https://github.com/software-mansion/scarb/pull/1570/commits/c0077eaf95f63b9f925b0538909b246bff4fe4d3